compiler/compiler.go

```diff
@@ -86,45 +86,180 @@ func compileFileDescriptorSets(files, importPaths []string, includeImports bool)
+		if err := resolveUninterpretedOptions(r, fd.GetOptions()); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func resolveMessageOptions(r resolver, md *pb.DescriptorProto) error {
+	if err := resolveUninterpretedOptions(r, md.GetOptions()); err != nil {
+		return err
+	}
+	for _, fd := range md.GetField() {
+		if err := resolveUninterpretedOptions(r, fd.GetOptions()); err != nil {
+			return err
+		}
+	}
+	for _, fd := range md.GetExtension() {
+		if err := resolveUninterpretedOptions(r, fd.GetOptions()); err != nil {
+			return err
+		}
+	}
+	for _, nestedMD := range md.GetNestedType() {
+		if err := resolveMessageOptions(r, nestedMD); err != nil {
+			return err
+		}
+	}
+	for _, ed := range md.GetEnumType() {
+		if err := resolveEnumOptions(r, ed); err != nil {
+			return err
+		}
+	}
+	for _, erd := range md.GetExtensionRange() {
+		if err := resolveUninterpretedOptions(r, erd.GetOptions()); err != nil {
+			return err
+		}
+	}
+	for _, od := range md.GetOneofDecl() {
+		if err := resolveUninterpretedOptions(r, od.GetOptions()); err != nil {
+			return err
 		}
 	}
 	return nil
 }
 
-func fullName(names []*pb.UninterpretedOption_NamePart) protoreflect.FullName {
-	// TODO: extend for field references and extensions of extensions.
-	str := protoreflect.FullName((*names[0].NamePart)[1:])
-	return str
+func resolveEnumOptions(r resolver, ed *pb.EnumDescriptorProto) error {
+	if err := resolveUninterpretedOptions(r, ed.GetOptions()); err != nil {
+		return err
+	}
+	for _, evd := range ed.GetValue() {
+		if err := resolveUninterpretedOptions(r, evd.GetOptions()); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func resolveServiceOptions(r resolver, sd *pb.ServiceDescriptorProto) error {
+	if err := resolveUninterpretedOptions(r, sd.GetOptions()); err != nil {
+		return err
+	}
+	for _, md := range sd.GetMethod() {
+		if err := resolveUninterpretedOptions(r, md.GetOptions()); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+type messageWithOptions interface {
+	proto.Message
+	GetUninterpretedOption() []*pb.UninterpretedOption
+}
+
+func resolveUninterpretedOptions(r resolver, opts messageWithOptions) error {
+	if opts == nil || reflect.ValueOf(opts).IsNil() {
+		return nil
+	}
+
+	for _, opt := range opts.GetUninterpretedOption() {
+		protoText := UninterpretedOptionAsProtoText(opt.Name, opt)
+		m := opts.ProtoReflect().New().Interface()
+		o := prototext.UnmarshalOptions{Resolver: r}
+		if err := o.Unmarshal([]byte(protoText), m); err != nil {
+			return err
+		}
+		proto.Merge(opts, m)
+	}
+
+	// Use reflection to set the UninterpretedOption field to nil
+	v := reflect.ValueOf(opts).Elem().FieldByName("UninterpretedOption")
+	if !v.IsZero() {
+		v.Set(reflect.Zero(v.Type()))
+	}
+
+	return nil
+}
+
+func UninterpretedOptionAsProtoText(names []*pb.UninterpretedOption_NamePart, val *pb.UninterpretedOption) string {
+	var result string
+	for i := len(names) - 1; i >= 0; i-- {
+		fieldName := *names[i].NamePart
```

Comment:
I assume these are messages, should we be using `GetNamePart()` et al here in case these somehow end up nil?